include/nbl/builtin/hlsl/bxdf/fresnel.hlsl

```diff
@@ -172,58 +240,53 @@ struct ReflectRefract
     using vector_type = vector<T, 3>;
     using scalar_type = T;
 
-    static this_t create(NBL_CONST_REF_ARG(vector_type) I, NBL_CONST_REF_ARG(vector_type) N, scalar_type NdotI, scalar_type rcpOrientedEta)
+    static this_t create(NBL_CONST_REF_ARG(vector_type) I, NBL_CONST_REF_ARG(vector_type) N, scalar_type rcpOrientedEta)
     {
         this_t retval;
-        retval.I = I;
-        retval.N = N;
-        retval.NdotI = NdotI;
+        retval._refract = Refract<scalar_type>::create(I, N);
         retval.rcpOrientedEta = rcpOrientedEta;
         return retval;
     }
 
     static this_t create(NBL_CONST_REF_ARG(Refract<scalar_type>) refract, scalar_type rcpOrientedEta)
     {
         this_t retval;
-        retval.I = refract.I;
-        retval.N = refract.N;
-        retval.NdotI = refract.NdotI;
+        retval._refract = refract;
         retval.rcpOrientedEta = rcpOrientedEta;
         return retval;
     }
 
     // when you know you'll reflect
-    void recomputeNdotR()
+    scalar_type getNdotR()
     {
-        _refract.recomputeNdotI();
-        NdotI = _refract.NdotI;
+        return _refract.getNdotI();
     }
 
     // when you know you'll refract
-    void recomputeNdotT(bool backside, scalar_type _NdotI2, scalar_type rcpOrientedEta2)
+    scalar_type getNdotT()
     {
-        _refract.recomputeNdotT(backside, _NdotI2, rcpOrientedEta2);
+        return _refract.getNdotT(rcpOrientedEta*rcpOrientedEta);
     }
 
-    vector_type reflect()
+    scalar_type getNdotTorR(const bool doRefract)
     {
-        return N * NdotI * 2.0f - I;
+        return hlsl::mix(getNdotR(), getNdotT(), doRefract);
     }
 
-    vector_type refract()
+    vector_type operator()(const bool doRefract)
     {
-        return N * (NdotI * rcpOrientedEta + _refract.NdotT) - I * rcpOrientedEta;
+        scalar_type NdotI = getNdotR();
+        return _refract.N * (NdotI * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract)) + hlsl::mix(NdotI, getNdotT(), doRefract)) - _refract.I * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract));
     }
 
-    vector_type operator()(const bool doRefract)
+    vector_type operator()(const scalar_type NdotTorR)
     {
-        return N * (NdotI * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract)) + hlsl::mix(NdotI, _refract.NdotT, doRefract)) - I * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract));
+        scalar_type NdotI = getNdotR();
+        bool doRefract = ComputeMicrofacetNormal<scalar_type>::isTransmissionPath(NdotI, NdotTorR);
+        return _refract.N * (NdotI * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract)) + hlsl::mix(NdotI, NdotTorR, doRefract)) - _refract.I * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract));
     }
 
     Refract<scalar_type> _refract;
-    vector_type I;
-    vector_type N;
-    scalar_type NdotI;
     scalar_type rcpOrientedEta;
```

Comment:
be consistent, either refract is allowed to keep the rcpOrientedEta, and then it lives in `_refract` for this struct, or both can't keep it